File: src/main/java/net/auoeke/limitless/Limitless.java
package net.auoeke.limitless;

import com.terraformersmc.modmenu.api.ConfigScreenFactory;
import com.terraformersmc.modmenu.api.ModMenuApi;
import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
import me.shedaniel.autoconfig.AutoConfig;
import me.shedaniel.autoconfig.serializer.GsonConfigSerializer;
import net.auoeke.limitless.config.Configuration;
import net.auoeke.limitless.config.enchantment.provider.EnchantmentListProvider;
import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.api.EnvironmentInterface;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.command.v1.CommandRegistrationCallback;
import net.minecraft.item.ItemStack;

@EnvironmentInterface(value = EnvType.CLIENT, itf = ClientModInitializer.class)
public class Limitless implements ModInitializer, ClientModInitializer {
    public static final String ID = "limitless";

    public static final ReferenceOpenHashSet<ItemStack> forConflictRemoval = new ReferenceOpenHashSet<>();

    @Override
    public void onInitialize() {
        Configuration.holder = AutoConfig.register(Configuration.class, GsonConfigSerializer::new);
        Configuration.refresh();
        CommandRegistrationCallback.EVENT.register(LimitlessCommand::register);
    }

    @Override
    @Environment(EnvType.CLIENT)
    public void onInitializeClient() {
        AutoConfig.getGuiRegistry(Configuration.class).registerPredicateProvider(new EnchantmentListProvider(), field -> field.getName().equals("maxLevels"));
    }

    @Environment(EnvType.CLIENT)
    @SuppressWarnings("unused")
    public static class ModMenu implements ModMenuApi {
        @Override
        public ConfigScreenFactory<?> getModConfigScreenFactory() {
            return parent -> AutoConfig.getConfigScreen(Configuration.class, parent).get();
        }
    }
}

File: src/net/auoeke/limitless/Limitless.kt
package net.auoeke.limitless

import com.terraformersmc.modmenu.api.ConfigScreenFactory
import com.terraformersmc.modmenu.api.ModMenuApi
import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet
import me.shedaniel.autoconfig.AutoConfig
import me.shedaniel.autoconfig.serializer.GsonConfigSerializer
import net.auoeke.limitless.config.Configuration
import net.auoeke.limitless.config.ExcludingNonprovider
import net.auoeke.limitless.config.enchantment.provider.EnchantmentListProvider
import net.fabricmc.api.*
import net.fabricmc.fabric.api.command.v1.CommandRegistrationCallback
import net.minecraft.item.ItemStack
import java.lang.reflect.Modifier

@EnvironmentInterface(value = EnvType.CLIENT, itf = ClientModInitializer::class)
object Limitless : ModInitializer, ClientModInitializer {
    const val ID: String = "limitless"
    val forConflictRemoval: ReferenceOpenHashSet<ItemStack> = ReferenceOpenHashSet()

    override fun onInitialize() {
        Configuration.holder = AutoConfig.register(Configuration::class.java) {definition, configClass -> GsonConfigSerializer(definition, configClass)}
        Configuration.refresh()
        CommandRegistrationCallback.EVENT.register(LimitlessCommand)
    }

    @Environment(EnvType.CLIENT)
    override fun onInitializeClient() {
        AutoConfig.getGuiRegistry(Configuration::class.java).apply {
            registerPredicateProvider(EnchantmentListProvider) {field -> field.name == "maxLevels"}
            registerPredicateProvider(ExcludingNonprovider) {field -> field.modifiers and (Modifier.STATIC or Modifier.TRANSIENT or 0x1000 /*synthetic*/) != 0}
        }
    }

    @Environment(EnvType.CLIENT)
    class ModMenu : ModMenuApi {
        override fun getModConfigScreenFactory(): ConfigScreenFactory<*> = ConfigScreenFactory {parent -> AutoConfig.getConfigScreen(Configuration::class.java, parent).get()}
    }
}
